### PR TITLE
Heartbeat Refactor

### PR DESCRIPTION
**Pull request by Lucien950**
### Changelist 

- Changed structure of heartbeat monitor config. (see `firmware/shared/src/app/app_heartbeatMonitorBoard.h`)
- Gives sturcture towards future app level refactor

### Testing Done

- fixed tests for all boards
- adapted shared heartbeat monitor to new structure

### CODE REVIEW

File: firmware/quadruna/BMS/src/app/app_heartbeatMonitor.c [hunk 1]
```diff
@@ -0,0 +1,53 @@
+#include "app_heartbeatMonitor.h"
+#include "app_heartbeatMonitorBoard.h"
+
+#include "app_canRx.h"
+#include "app_canTx.h"
+#include "app_canAlerts.h"
+
+HeartbeatMonitorBoard vc_hbmonitor = {
```
Comment (liam-ilan) [inline firmware/quadruna/BMS/src/app/app_heartbeatMonitor.c:8]: Nit: This should probably be named something more consistent with the naming scheme on the rest of the codebase (`VC_heartbeat_monitor` or `VC_hb_monitor`?)

File: firmware/quadruna/BMS/src/app/app_heartbeatMonitor.c [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+void app_heartbeatMonitor_init(bool block_faults_init)
+{
+    block_faults = block_faults_init;
+    app_heartbeatMonitorBoard_init(&vc_hbmonitor);
+}
+
+void app_heartbeatMonitor_checkIn(void)
+{
+    app_canTx_BMS_Heartbeat_set(true);
```
Comment (liam-ilan) [inline firmware/quadruna/BMS/src/app/app_heartbeatMonitor.c:26]: We can pass the heartbeat setter function to `HeartbeatMonitorBoard`, so that the hb monitor library can handle this step - then we can expose a method like `app_heartbeatMonitorBoard_tick` rather than `app_heartbeatMonitorBoard_checkIn`...

I'm just thinking that for the next dev using the heartbeat monitor, it seems like a reasonable mistake to forget to broadcast the heartbeat manually.

File: firmware/quadruna/BMS/src/app/app_heartbeatMonitor.c [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    app_heartbeatMonitorBoard_checkIn(&vc_hbmonitor);
+}
+
+void app_heartbeatMonitor_broadcastFaults(void)
+{
+    if (block_faults)
```
Comment (liam-ilan) [inline firmware/quadruna/BMS/src/app/app_heartbeatMonitor.c:32]: fault blocking can also be set as a flag in the `HeartbeatMonitorBoard` struct if we want the end developer to use a flag to disable faults rather than manually checking it.

File: firmware/quadruna/BMS/src/app/app_heartbeatMonitor.c [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+bool app_heartbeatMonitor_isSendingMissingHeartbeatFault(void)
+{
+    if (block_faults)
+        return false;
+    return app_heartbeatMonitorBoard_isSendingMissingHeartbeatFault(&vc_hbmonitor);
+}
+
+#ifdef TARGET_TEST
+void app_heartbeatMonitor_clearFaults(void)
```
Comment (liam-ilan) [inline firmware/quadruna/BMS/src/app/app_heartbeatMonitor.c:45]: Can we put this target check in the heartbeat monitor library?

File: firmware/quadruna/BMS/src/app/app_heartbeatMonitor.c [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+{
+    app_heartbeatMonitorBoard_clearFaults(&vc_hbmonitor);
+}
+void app_heartbeatMonitor_blockFaults(bool block)
+{
+    block_faults = block;
+}
+#endif
```
Comment (liam-ilan) [inline firmware/quadruna/BMS/src/app/app_heartbeatMonitor.c:53]: I think that we can move a lot of the logic in this intermediate heartbeat layer to the heartbeat monitor library, that way the only thing left here will be configuration. Should help the next dev when they need to port a board (this looks like a site for a lot of potential mistakes).

File: firmware/quadruna/CRIT/src/app/app_heartbeatMonitor.c [hunk 6]
```diff
@@ -0,0 +1,84 @@
+#include "app_heartbeatMonitor.h"
```
Comment (liam-ilan) [inline firmware/quadruna/CRIT/src/app/app_heartbeatMonitor.c:1]: Same notes from BMS apply here - I think we can move any remaining logic in this file and move it to the heartbeat monitor library, making this just a configuration file.

File: firmware/quadruna/CRIT/src/app/app_heartbeatMonitor.c [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+HeartbeatMonitorBoard fsm_hbmonitor = { .getter       = app_canRx_FSM_Heartbeat_get,
+                                        .resetter     = app_canRx_FSM_Heartbeat_update,
+                                        .fault_setter = app_canAlerts_CRIT_Fault_MissingFSMHeartbeat_set,
+                                        .fault_getter = app_canAlerts_CRIT_Fault_MissingFSMHeartbeat_get,
+                                        .timeout_ms   = 200 };
+
+static bool block_faults = false;
+
+void app_heartbeatMonitor_init(bool block_faults_init)
```
Comment (liam-ilan) [inline firmware/quadruna/CRIT/src/app/app_heartbeatMonitor.c:34]: The init wrapper here is nice though - we should keep this method here.

**Review by liam-ilan — CHANGES_REQUESTED** [review]
Overall a huge improvement over the configuration blocks we had earlier - way cleaner. Just some minor notes about the intermediate API layer that we have on every board that wants to use the heartbeat monitor. Other than that lets get this money!

File: firmware/cmake/tests.cmake [hunk 8]
```diff
@@ -23,7 +23,6 @@ function(compile_gtest_executable
     target_compile_options(${TEST_EXECUTABLE_NAME}
         PUBLIC
         -Wall
-        -g3
```
Comment (liam-ilan) [inline firmware/cmake/tests.cmake:26]: Any reason why we drop `-g3`?

File: firmware/logfs/CMakeLists.txt [hunk 9]
```diff
@@ -1,3 +1,6 @@
+message("")
+message("⚙️ Configuring LogFS")
```
Comment (liam-ilan) [inline firmware/logfs/CMakeLists.txt:2]: Legit to leave this here - just checking that this isn't changes from another PR leaking into this one - if it isn't ignore this comment 😅

File: firmware/quadruna/BMS/src/app/app_heartbeatMonitors.c [hunk 10]
```diff
@@ -0,0 +1,24 @@
+//
+// Created by jessi on 2024-12-28.
+//
+
+#include "app_heartbeatMonitors.h"
+#include "app_canRx.h"
+#include "app_canAlerts.h"
+#include "app_canTx.h"
+
+static HeartbeatMonitorBoard boards[1] = {
```
Comment (liam-ilan) [inline firmware/quadruna/BMS/src/app/app_heartbeatMonitors.c:10]: This is excellent - so much better than the the previous refactor. One nit though - can we call this `heartbeat_boards`?

File: firmware/quadruna/BMS/src/app/app_heartbeatMonitors.c [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        .getter       = app_canRx_VC_Heartbeat_get,
+        .resetter     = app_canRx_VC_Heartbeat_update,
+        .fault_setter = app_canAlerts_BMS_Warning_MissingVCHeartbeat_set,
+        .fault_getter = app_canAlerts_BMS_Warning_MissingVCHeartbeat_get,
+        .timeout_ms   = 200,
+    }
+};
+
+HeartbeatMonitor BMS_heartbeat_monitor = { .boards           = boards,
```
Comment (liam-ilan) [inline firmware/quadruna/BMS/src/app/app_heartbeatMonitors.c:21]: Nit: No need to specify `BMS` here.

File: firmware/quadruna/BMS/src/app/app_heartbeatMonitors.c [hunk 12]
```diff
@@ -0,0 +1,24 @@
+//
+// Created by jessi on 2024-12-28.
```
Comment (liam-ilan) [inline firmware/quadruna/BMS/src/app/app_heartbeatMonitors.c:2]: Who is Jessi?

File: firmware/quadruna/BMS/src/app/app_heartbeatMonitors.h [hunk 13]
```diff
@@ -0,0 +1,5 @@
+#pragma once
+
+#include "app_heartbeatMonitor.h"
+
+extern HeartbeatMonitor BMS_heartbeat_monitor;
```
Comment (liam-ilan) [inline firmware/quadruna/BMS/src/app/app_heartbeatMonitors.h:5]: I like this - I like this a lot. Nit though: Let's rename this to `app_heartbeatMonitor_cfg.h` or something along that line... While this file is technically a collection of heartbeat monitors, it's more informative to tell the developer that this is where the heartbeat monitor is configured.

File: firmware/quadruna/BMS/src/tasks.c [hunk 14]
```diff
@@ -211,41 +211,6 @@ static const GlobalsConfig globals_config = { .bms_ok_latch  = &bms_ok_latch,
...
-    [DIM_HEARTBEAT_BOARD] = NULL, [CRIT_HEARTBEAT_BOARD] = NULL
-};
-
-// heartbeatFaultGetters - gets fault statuses over CAN
-static bool (*const heartbeatFaultGetters[HEARTBEAT_BOARD_COUNT])(void) = {
-    [BMS_HEARTBEAT_BOARD] = NULL, [VC_HEARTBEAT_BOARD] = app_canAlerts_BMS_Warning_MissingVCHeartbeat_get,
-    [RSM_HEARTBEAT_BOARD] = NULL, [FSM_HEARTBEAT_BOARD] = NULL,
-    [DIM_HEARTBEAT_BOARD] = NULL, [CRIT_HEARTBEAT_BOARD] = NULL
-};
-
```
Comment (liam-ilan) [inline firmware/quadruna/BMS/src/tasks.c:248]: Beautiful.

File: firmware/quadruna/BMS/test/test_imd.h [hunk 15]
```diff
@@ -22,14 +22,14 @@ class ImdTest : public testing::Test
...
 {
-    const float mapping[NUM_OF_IMD_CONDITIONS] = {
-        [IMD_CONDITION_SHORT_CIRCUIT] = 0.0f,          [IMD_CONDITION_NORMAL] = 10.0f,
-        [IMD_CONDITION_UNDERVOLTAGE_DETECTED] = 20.0f, [IMD_CONDITION_SST] = 30.0f,
-        [IMD_CONDITION_DEVICE_ERROR] = 40.0f,          [IMD_CONDITION_GROUND_FAULT] = 50.0f,
+    const std::map<ImdConditionName, float> mapping{
+        { IMD_CONDITION_SHORT_CIRCUIT, 0.0f },          { IMD_CONDITION_NORMAL, 10.0f },
+        { IMD_CONDITION_UNDERVOLTAGE_DETECTED, 20.0f }, { IMD_CONDITION_SST, 30.0f },
+        { IMD_CONDITION_DEVICE_ERROR, 40.0f },          { IMD_CONDITION_GROUND_FAULT, 50.0f }
     };
```
Comment (liam-ilan) [inline firmware/quadruna/BMS/test/test_imd.h:31]: Again - just making sure this isn't a leak from another PR/if you maybe want to split it to another PR... I understand the appeal of just getting this in so I'm cool with that, just lmk

File: firmware/shared/src/app/app_heartbeatMonitor.h [hunk 16]
```diff
@@ -1,86 +1,22 @@
...
-    // FAULT SETGET
-
-    // fault broadcasters for each board's heartbeat from this board
-    void (*fault_setters[HEARTBEAT_BOARD_COUNT])(bool);
-    // fault getters for each board's heartbeat from this board
-    bool (*fault_getters[HEARTBEAT_BOARD_COUNT])(void);
+    HeartbeatMonitorBoard *boards;
+    uint8_t                board_count;
+    bool                   block_faults;
+    void (*self_heartbeater)(bool);
```
Comment (liam-ilan) [inline firmware/shared/src/app/app_heartbeatMonitor.h:10]: Nit: how about `self_heartbeat` or `own_heartbeat` or `heartbeat_broadcaster` or `heartbeat_tx` or something like that? 😅

**Review by liam-ilan — CHANGES_REQUESTED** [review]
Some nits - Commented on the BMS implementation but I think most of the comments apply across all boards, ping me when you want another review!

File: firmware/cmake/tests.cmake (hunk 8 above)
Comment (Lucien950) [reply firmware/cmake/tests.cmake:26]: redundant because it is specified with CMAKE_BUILD_TYPE.

the idea is to support as many toolchains as possible. -g3 is gcc flag only, /g3 for msvc for example.

File: firmware/logfs/CMakeLists.txt (hunk 9 above)
Comment (Lucien950) [reply firmware/logfs/CMakeLists.txt:2]: oh i was just debugging

File: firmware/quadruna/BMS/src/app/app_heartbeatMonitors.c (hunk 12 above)
Comment (Lucien950) [reply firmware/quadruna/BMS/src/app/app_heartbeatMonitors.c:2]: me gang 😭

File: firmware/quadruna/BMS/src/app/app_heartbeatMonitors.h (hunk 13 above)
Comment (Lucien950) [reply firmware/quadruna/BMS/src/app/app_heartbeatMonitors.h:5]: I think this gets into a question of what to do with all the board specific config files (very much including hardware level stuff). I just named it this way to keep consistent with the way all the other shared library using modules do it.

also i think atm this is just a reference implementation, and we can figure out in the new year what to do with these types of files

File: firmware/quadruna/BMS/test/test_imd.h (hunk 15 above)
Comment (Lucien950) [reply firmware/quadruna/BMS/test/test_imd.h:31]: was breaking on my computer

File: firmware/quadruna/BMS/src/tasks.c (hunk 14 above)
Comment (Lucien950) [reply firmware/quadruna/BMS/src/tasks.c:248]: :)

File: firmware/quadruna/BMS/src/app/app_heartbeatMonitors.h (hunk 13 above)
Comment (liam-ilan) [reply firmware/quadruna/BMS/src/app/app_heartbeatMonitors.h]: I'm cool with merging in this state to move things forward then

File: firmware/quadruna/BMS/test/test_imd.h (hunk 15 above)
Comment (liam-ilan) [reply firmware/quadruna/BMS/test/test_imd.h:31]: Interesting - cool cool!

**Review by liam-ilan — APPROVED** [review]
LGTM
